README.md
# nvd3-charts
Custom charts for nvd3.

By Johan Johansson/Enplore (https://github.com/enplore/nvd3-charts.git).

Current version checked OK with nvd3 1.8.4 and d3 3.5.17.

## Install
### Manually
Download the scripts and include them after including d3 and nvd3 scripts. To use, just use nv.models.xxx as per below.

### With npm
The repo is not published on npmjs, however you can still install it with npm from github:

`npm install --save github:enplore/nvd3-charts`

Then `import 'nvd3-charts/gaugeChart';` after importing d3 and nvd3 packages. To use, just use nv.models.xxx as per below.

## Use
### Gauge chart
Simple gauge, wants a single value as data. Use CSS widths to adjust size (not responsive).

Live demo: http://codepen.io/anon/pen/OXprXo

![](https://raw.githubusercontent.com/enplore/nvd3-charts/master/gauge.png)

```
nv.addGraph(function() {
    var chart = nv.models.gaugeChart()
        .title('Gauge')
        .min(0)
        .max(1)
        .zoneLimit1(0.25)
        .zoneLimit2(0.75);

    d3.select('#gauge-chart svg')
        .datum([Math.random()])
        .call(chart);

    nv.utils.windowResize(chart.update);
    return chart;
});
```

### Packed bubble chart (hierarchical)
Based on http://bl.ocks.org/mbostock/4063269. Accepts a similar data structure (see demo for exact structure). Responsive. Will reveal child node when clicked.

Live demo: http://codepen.io/anon/pen/yJMGVB

```
nv.addGraph(function() {
    var chart = nv.models.packedBubbleChart()
        .title('Bubbles')
        .valueFormat(function(d) { return d.index.toFixed(2); })
        .color(d3.scale.linear()
              .domain([0, 0.25, 0.5])
              .range(['#88ac67', "#f78f20", "#db4e4e"]));

    d3.select('#bubble-chart svg')
        .datum(bubbleData)
        .call(chart);

    nv.utils.windowResize(chart.update);
    return chart;
});
```

### Radar chart
Based on http://bl.ocks.org/nbremer/6506614. Accepts a similar data structure (see demo for exact structure). Nodes are linked if the data point's "link" property is set.

Live demo: http://codepen.io/anon/pen/GqWPyX

```
nv.addGraph(function() {
    var chart = nv.models.radarChart()
        .valueFormat(function (d) { return d.toFixed(2); })
        .min(min)
        .max(max)
        .stepSize(tickStep)
        .margin({ top: 10 })
        .color(d3.scale.linear()
            .range(['#88ac67', "#f78f20", "#db4e4e"]));

    d3.select('#radar-chart svg')
        .datum(radarData)
        .call(chart);

    nv.utils.windowResize(chart.update);
    return chart;
});
```


CHANGELOG.md
# Change Log
## [0.0.6] - 2017-03-15
- Fixed needle no longer pointing passed the maximum value.

## [0.0.5] - 2017-02-07
- Added zone limit variables to allow custom color ranges.

## [0.0.3] - 2016-03-23
- Fixed tooltip positioning in radar chart with nvd3 1.8.4.

## [0.0.3] - 2016-03-23
- Fixed tooltip positioning in bubble chart with nvd3 1.8.2.

## [0.0.2] - 2016-01-21
- Added step size to radar chart.


gauge.js
nv.models.gauge = function() {
    "use strict";

    //============================================================
    // Public Variables with Default Settings
    //------------------------------------------------------------

    var margin = {top: 0, right: 0, bottom: 0, left: 0}
        , width = 500
        , height = 500
        , id = Math.floor(Math.random() * 10000) //Create semi-unique ID in case user doesn't select one
        , color = nv.utils.getColor(['#88ac67', '#f78f20', '#db4e4e'])
        , valueFormat = d3.format(',.2f')
        , title = false
        , showMinMaxLabels = false
        , min = 0
        , max = 100
        , zoneLimit1 = 0.6
        , zoneLimit2 = 0.8
        , dispatch = d3.dispatch('chartClick', 'renderEnd')
        ;


    //============================================================
    // chart function
    //------------------------------------------------------------

    var renderWatch = nv.utils.renderWatch(dispatch);

    function chart(selection) {
        renderWatch.reset();

        selection.each(function(data) {
            var availableWidth = width - margin.left - margin.right
                , availableHeight = height - margin.top - margin.bottom
                , container = d3.select(this)
                ;

            var cx = availableWidth / 2;
            var cy = availableHeight / 2;

            nv.utils.initSVG(container);

            var radius = Math.min(availableWidth, availableHeight) / 2;
            var range = max - min;
            var fontSize = Math.round(Math.min(availableWidth, availableHeight) / 9);

            var zones = [
                { from: min, to: min + range * zoneLimit1 },
                { from: min + range * zoneLimit1, to: min + range * zoneLimit2 },
                { from: min + range * zoneLimit2, to: max }
            ];

            // Setup containers and skeleton of chart
            var wrap = container.selectAll('.nv-wrap.nv-gauge').data([data]);
            var wrapEnter = wrap.enter().append('g').attr('class','nvd3 nv-wrap nv-gauge nv-chart-' + id);
            var gEnter = wrapEnter.append('g');
            var g_bands = gEnter.append('g').attr('class', 'nv-gaugeBands');
            var g_title = gEnter.append('g').attr('class', 'nv-gaugeTitle');
            var g_needle = gEnter.append('g').attr('class', 'nv-gaugeNeedle');
            var g_label = gEnter.append('g').attr('class', 'nv-gaugeLabel');
            var g_minLabel = gEnter.append('g').attr('class', 'nv-gaugeMinLabel');
            var g_maxLabel = gEnter.append('g').attr('class', 'nv-gaugeMaxLabel');

            wrap.attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');

            // draw gauge bands
            for (var i in zones) {
                drawBand(zones[i].from, zones[i].to, color(i), min, max, radius, g_bands);
            }

            // draw needle
            var needlePath = buildNeedlePath(data[0][0], range, cx, cy, min, max, radius);

            var needleLine = d3.svg.line()
                .x(function(d) { return d.x; })
                .y(function(d) { return d.y; })
                .interpolate("basis");

            g_needle.append("path")
                .data([needlePath])
                .attr("d", needleLine);

            g_needle.append('circle')
                .attr('cx', 0)
                .attr('cy', 0)
                .attr('r', 0.115 * radius);

            wrap.selectAll('.nv-gaugeBands path')
                .attr("transform", function () { return "translate(" + cx + ", " + (cy - 0.08 * radius) + ") rotate(270)" });

            wrap.select('.nv-gaugeNeedle')
                .attr('transform', 'translate(' + cx + ',' + (cy - 0.08 * radius) + ')');

            wrap.select('.nv-gaugeTitle')
                .attr('transform', 'translate(' + cx + ',' + (cy / 2 + fontSize / 2) + ')');

            wrap.select('.nv-gaugeLabel')
                .attr('transform', 'translate(' + cx + ',' + (cy + radius / 2 - fontSize * 0.9) + ')');

            if (showMinMaxLabels) {
                wrap.select('.nv-gaugeMinLabel')
                    .attr('transform', 'translate(' + (cx - radius / 2.6 - fontSize * 0.9) + ',' + (cy + radius / 1.35 - fontSize * 0.9) + ')');

                wrap.select('.nv-gaugeMaxLabel')
                    .attr('transform', 'translate(' + (cx + radius / 1.25 - fontSize * 0.9) + ',' + (cy + radius / 1.35 - fontSize * 0.9) + ')');
            }

            // draw title
            if (title) {
                g_title.append("text")
                    .attr("dy", fontSize / 2)
                    .attr("text-anchor", "middle")
                    .text(title)
                    .style("font-size", fontSize + "px");
            }

            // draw value
            g_label.append("text")
                .data(data)
                .attr("dy", fontSize / 2)
                .attr("text-anchor", "middle")
                .text(valueFormat)
                .style("font-size", fontSize*0.9 + "px");

            if (showMinMaxLabels) {
                g_minLabel.append("text")
                    .data(data)
                    .attr("dy", fontSize / 2)
                    .attr("text-anchor", "start")
                    .text(valueFormat(min))
                    .style("font-size", fontSize*0.45 + "px");

                g_maxLabel.append("text")
                    .data(data)
                    .attr("dy", fontSize / 2)
                    .attr("text-anchor", "end")
                    .text(valueFormat(max))
                    .style("font-size", fontSize*0.45 + "px");
            }

            container.on('click', function(d,i) {
                dispatch.chartClick({
                    data: d,
                    index: i,
                    pos: d3.event,
                    id: id
                });
            });

            // draws a gauge band
            function drawBand(start, end, color, min, max, radius, element) {
                if (0 >= end - start) return;

                element.append("path")
                    .style("fill", color)
                    .attr("d", d3.svg.arc()
                        .startAngle(valueToRadians(start, min, max))
                        .endAngle(valueToRadians(end, min, max))
                        .innerRadius(0.65 * radius)
                        .outerRadius(0.85 * radius))
                    .attr("transform", function() { return "translate(" + radius + ", " + radius + ") rotate(270)" });
            }

            function buildNeedlePath(value, range, cx, cy, min, max, radius) {
                if (value > max) {
                    value = max;
                }

                var delta = range / 1;
                var tailValue = value - (range * (1/(270/360)) / 2);

                var head = centerPoint(valueToPoint(value, 0.8, min, max, radius), cx, cy);
                var head1 = centerPoint(valueToPoint(value - delta, 0.12, min, max, radius), cx, cy);
                var head2 = centerPoint(valueToPoint(value + delta, 0.12, min, max, radius), cx, cy);

                var tail = centerPoint(valueToPoint(tailValue, 0, min, max, radius), cx, cy);
                var tail1 = centerPoint(valueToPoint(tailValue - delta, 0.12, min, max, radius), cx, cy);
                var tail2 = centerPoint(valueToPoint(tailValue + delta, 0.12, min, max, radius), cx, cy);

                function centerPoint(point, cx, cy) {
                    point.x -= cx;
                    point.y -= cy;
                    return point;
                }

                return [head, head1, tail2, tail, tail1, head2, head];
            }

            function valueToDegrees(value, min, max) {
                range = max - min;
                return value / range * 270 - (min / range * 270 + 45);
            }

            function valueToRadians(value, min, max) {
                return valueToDegrees(value, min, max) * Math.PI / 180;
            }

            function valueToPoint(value, factor, min, max, radius) {
                return {
                    x: cx - radius * factor * Math.cos(valueToRadians(value, min, max)),
                    y: cy - radius * factor * Math.sin(valueToRadians(value, min, max))
                };
            }
        });

        renderWatch.renderEnd('gauge immediate');
        return chart;
    }

    //============================================================
    // Expose Public Variables
    //------------------------------------------------------------

    chart.dispatch = dispatch;
    chart.options = nv.utils.optionsFunc.bind(chart);

    chart._options = Object.create({}, {
        // simple options, just get/set the necessary values
        width:      {get: function(){return width;}, set: function(_){width=_;}},
        height:     {get: function(){return height;}, set: function(_){height=_;}},
        title:      {get: function(){return title;}, set: function(_){title=_;}},
        showMinMaxLabels:    {get: function(){return showMinMaxLabels;}, set: function(_){showMinMaxLabels=_;}},
        valueFormat:    {get: function(){return valueFormat;}, set: function(_){valueFormat=_;}},
        id:         {get: function(){return id;}, set: function(_){id=_;}},
        min:         {get: function(){return min;}, set: function(_){min=_;}},
        max:         {get: function(){return max;}, set: function(_){max=_;}},
        zoneLimit1: {get: function(){return zoneLimit1;}, set: function(_){zoneLimit1=_;}},
        zoneLimit2: {get: function(){return zoneLimit2;}, set: function(_){zoneLimit2=_;}},

        // options that require extra logic in the setter
        margin: {get: function(){return margin;}, set: function(_){
            margin.top    = typeof _.top    != 'undefined' ? _.top    : margin.top;
            margin.right  = typeof _.right  != 'undefined' ? _.right  : margin.right;
            margin.bottom = typeof _.bottom != 'undefined' ? _.bottom : margin.bottom;
            margin.left   = typeof _.left   != 'undefined' ? _.left   : margin.left;
        }},
        color: {get: function(){return color;}, set: function(_){
            color=nv.utils.getColor(_);
        }}
    });

    nv.utils.initOptions(chart);
    return chart;
};

gaugeChart.js
nv.models.gaugeChart = function() {
    "use strict";

    //============================================================
    // Public Variables with Default Settings
    //------------------------------------------------------------

    var gauge = nv.models.gauge();

    var margin = {top: 0, right: 0, bottom: 0, left: 0}
        , width = null
        , height = null
        , color = nv.utils.defaultColor()
        , noData = null
        , dispatch = d3.dispatch('renderEnd')
        ;

    //============================================================
    // Chart function
    //------------------------------------------------------------

    var renderWatch = nv.utils.renderWatch(dispatch);

    function chart(selection) {
        renderWatch.reset();
        renderWatch.models(gauge);

        selection.each(function(data) {
            var container = d3.select(this);
            nv.utils.initSVG(container);

            var availableWidth = nv.utils.availableWidth(width, container, margin),
                availableHeight = nv.utils.availableHeight(height, container, margin);

            chart.update = function() { container.transition().call(chart); };
            chart.container = this;

            // Display No Data message if there's nothing to show.
            if (!data || !data.length) {
                nv.utils.noData(chart, container);
                return chart;
            } else {
                container.selectAll('.nv-noData').remove();
            }

            // Setup containers and skeleton of chart
            var wrap = container.selectAll('g.nv-wrap.nv-gaugeChart').data([data]);
            var gEnter = wrap.enter().append('g').attr('class', 'nvd3 nv-wrap nv-gaugeChart').append('g');
            var g = wrap.select('g');

            gEnter.append('g').attr('class', 'nv-gaugeWrap');

            wrap.attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');

            // Main Chart Component(s)
            gauge.width(availableWidth).height(availableHeight);
            var gaugeWrap = g.select('.nv-gaugeWrap').datum([data]);
            d3.transition(gaugeWrap).call(gauge);
        });

        renderWatch.renderEnd('gauge chart immediate');
        return chart;
    }

    //============================================================
    // Expose Public Variables
    //------------------------------------------------------------

    // expose chart's sub-components
    chart.dispatch = dispatch;
    chart.gauge = gauge;
    chart.options = nv.utils.optionsFunc.bind(chart);

    // use Object get/set functionality to map between vars and chart functions
    chart._options = Object.create({}, {
        // simple options, just get/set the necessary values
        width:      {get: function(){return width;}, set: function(_){width=_;}},
        height:     {get: function(){return height;}, set: function(_){height=_;}},
        noData:         {get: function(){return noData;}, set: function(_){noData=_;}},

        // options that require extra logic in the setter
        color: {get: function(){return color;}, set: function(_){
            color = _;
            gauge.color(color);
        }},
        margin: {get: function(){return margin;}, set: function(_){
            margin.top    = _.top    !== undefined ? _.top    : margin.top;
            margin.right  = _.right  !== undefined ? _.right  : margin.right;
            margin.bottom = _.bottom !== undefined ? _.bottom : margin.bottom;
            margin.left   = _.left   !== undefined ? _.left   : margin.left;
        }}
    });

    nv.utils.inheritOptions(chart, gauge);
    nv.utils.initOptions(chart);

    return chart;
};


packedBubbleChart.js
nv.models.packedBubbleChart = function() {
    "use strict";

    //============================================================
    // Public Variables with Default Settings
    //------------------------------------------------------------

    var packedBubble = nv.models.packedBubble()
        , tooltip = nv.models.tooltip()
        ;

    var margin = {top: 0, right: 0, bottom: 0, left: 0}
        , width = null
        , height = null
        , color = nv.utils.defaultColor()
        , noData = null
        , dispatch = d3.dispatch('renderEnd')
        ;

    tooltip
        .headerEnabled(false)
        .duration(0)
        .valueFormatter(function(d, i) {
            return packedBubble.valueFormat()(d, i);
        });

    //============================================================
    // Chart function
    //------------------------------------------------------------

    var renderWatch = nv.utils.renderWatch(dispatch);

    function chart(selection) {
        renderWatch.reset();
        renderWatch.models(packedBubble);

        selection.each(function(data) {
            var container = d3.select(this);
            nv.utils.initSVG(container);

            var availableWidth = nv.utils.availableWidth(width, container, margin),
                availableHeight = nv.utils.availableHeight(height, container, margin);

            chart.update = function() { container.transition().call(chart); };
            chart.container = this;

            // Display No Data message if there's nothing to show.
            if (!data) {
                nv.utils.noData(chart, container);
                return chart;
            } else {
                container.selectAll('.nv-noData').remove();
            }

            // Setup containers and skeleton of chart
            var wrap = container.selectAll('g.nv-wrap.nv-packedBubbleChart').data([data]);
            var gEnter = wrap.enter().append('g').attr('class', 'nvd3 nv-wrap nv-packedBubbleChart').append('g');
            var g = wrap.select('g');

            gEnter.append('g').attr('class', 'nv-packedBubbleWrap');

            wrap.attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');

            // Main Chart Component(s)
            packedBubble.width(availableWidth).height(availableHeight);
            var packedBubbleWrap = g.select('.nv-packedBubbleWrap').datum(data);
            d3.transition(packedBubbleWrap).call(packedBubble);
        });

        renderWatch.renderEnd('packedBubble chart immediate');
        return chart;
    }

    //============================================================
    // Event Handling/Dispatching (out of chart's scope)
    //------------------------------------------------------------
    packedBubble.dispatch.on('elementMouseover.tooltip', function(evt) {
        evt['series'] = {
            key: evt.data.name,
            value: evt.data,
            color: evt.color
        };

        tooltip.data(evt).hidden(false);
    });

    packedBubble.dispatch.on('elementMouseout.tooltip', function(evt) {
        tooltip.hidden(true);
    });

    packedBubble.dispatch.on('elementMousemove.tooltip', function(evt) {
        tooltip.position();
    });

    //============================================================
    // Expose Public Variables
    //------------------------------------------------------------

    // expose chart's sub-components
    chart.dispatch = dispatch;
    chart.packedBubble = packedBubble;
    chart.options = nv.utils.optionsFunc.bind(chart);
    chart.tooltip = tooltip;

    // use Object get/set functionality to map between vars and chart functions
    chart._options = Object.create({}, {
        // simple options, just get/set the necessary values
        width:      {get: function(){return width;}, set: function(_){width=_;}},
        height:     {get: function(){return height;}, set: function(_){height=_;}},
        noData:         {get: function(){return noData;}, set: function(_){noData=_;}},

        // options that require extra logic in the setter
        color: {get: function(){return color;}, set: function(_){
            color = _;
            packedBubble.color(color);
        }},
        margin: {get: function(){return margin;}, set: function(_){
            margin.top    = _.top    !== undefined ? _.top    : margin.top;
            margin.right  = _.right  !== undefined ? _.right  : margin.right;
            margin.bottom = _.bottom !== undefined ? _.bottom : margin.bottom;
            margin.left   = _.left   !== undefined ? _.left   : margin.left;
        }}
    });

    nv.utils.inheritOptions(chart, packedBubble);
    nv.utils.initOptions(chart);

    return chart;
};


radar.js
nv.models.radar = function() {
    "use strict";

    //============================================================
    // Public Variables with Default Settings
    //------------------------------------------------------------

    var margin = {top: 0, right: 0, bottom: 0, left: 0}
        , width = 500
        , height = 500
        , id = Math.floor(Math.random() * 10000) //Create semi-unique ID in case user doesn't select one
        , color = nv.utils.getColor(['#88ac67', '#f78f20', '#db4e4e'])
        , valueFormat = d3.format(',.2f')
        , margin = { top: 15, bottom: 30, left: 100, right: 120 }
        , title = false
        , min = null
        , max = null
        , radius = 5
        , factor = 1
        , factorLegend = 1
        , levels = 10
        , opacityArea = 0.5
        , nodeRadius = 4
        , dispatch = d3.dispatch('chartClick', 'renderEnd', 'elementMouseover', 'elementMouseout', 'elementMousemove')
        ;


    //============================================================
    // chart function
    //------------------------------------------------------------

    var renderWatch = nv.utils.renderWatch(dispatch);

    function chart(selection) {
        renderWatch.reset();

        selection.each(function (data) {
            var availableWidth = width - margin.left - margin.right
                , availableHeight = height - margin.top - margin.bottom
                , container = d3.select(this)
                ;

            if (max == null) {
                data.forEach(function (d) {
                    var m = d3.max(d.values, function (v) { return v.value; });

                    if (m > max)
                        max = m;
                });
            }

            if (min == null) {
                min = -max;
            }

            radius = factor * (Math.min(availableWidth, availableHeight) / 2);

            data.forEach(function (d, i) {
                d.color = color(d.series);
            });

            nv.utils.initSVG(container);

            var step = (max - min) / levels;
            var adjustedMax = max + step;
            var range = d3.range(min, adjustedMax, step);
            var allAxis = data[0].values.map(function (d) { return d.axis; });
            var radians = 2 * Math.PI;

            // Setup containers and skeleton of chart
            var wrap = container.selectAll('.nv-wrap.nv-radar').data([data]);
            var wrapEnter = wrap.enter().append('g').attr('class','nvd3 nv-wrap nv-radar nv-chart-' + id);
            var gEnter = wrapEnter.append('g');

            wrap.attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');

            // level lines (circle segments)
            gEnter.append('g').attr('class', 'nv-radarLevels');
            var g_levels = wrap.select('.nv-radarLevels');

            var levelFactors = range.map(function (i, j) {
                return factor * radius * (j / (range.length - 1));
            });

            var levelGroups = g_levels
                .selectAll('.level-group')
                .data(levelFactors);

            levelGroups
                .enter()
                .append('g')
                .attr('class', 'level-group');

            levelGroups.exit().remove();

            var levelLines = levelGroups
                .selectAll('.level')
                .data(function (levelFactor) {
                    return allAxis.map(function (i) { return levelFactor; });
                });

            levelLines
                .enter()
                .append('line')
                .attr('class', 'level')
                .style("stroke", "grey")
                .style("stroke-opacity", "0.75")
                .style("stroke-width", "0.3px");

            levelLines.exit().remove();

            levelGroups.selectAll('.level')
                .attr('x1', function (levelFactor, i) { return levelFactor * (1 - factor * Math.sin(i * radians / allAxis.length)); })
                .attr("y1", function (levelFactor, i) { return levelFactor * (1 - factor * Math.cos(i * radians / allAxis.length)); })
                .attr("x2", function (levelFactor, i) { return levelFactor * (1 - factor * Math.sin((i + 1) * radians / allAxis.length)); })
                .attr("y2", function (levelFactor, i) { return levelFactor * (1 - factor * Math.cos((i + 1) * radians / allAxis.length)); })
                .attr("transform", function (levelFactor) { return "translate(" + (availableWidth / 2 - levelFactor) + ", " + (availableHeight / 2 - levelFactor) + ")"; });

            // level values
            gEnter.append('g').attr('class', 'nv-radarLevelValues');
            var g_labelValues = wrap.select('.nv-radarLevelValues');

            var labelValues = g_labelValues
                .selectAll('.level-value')
                .data(levelFactors);

            labelValues
                .enter()
                .append('text')
                .attr('class', 'level-value')
                .style('fill', '#000000')
                .text(function (d, i) { return range[i]; });

            labelValues.exit().remove();

            g_labelValues
                .selectAll('.level-value')
                .attr('transform', function (levelFactor, i) { return 'translate(' + (5 + availableWidth / 2) + ',' + (availableHeight / 2 - levelFactor) + ')'; });

            // axis lines
            gEnter.append('g').attr('class', 'nv-radarAxisLines');
            var g_axisLines = wrap.select('.nv-radarAxisLines');

            var axisLines = g_axisLines
                .selectAll('.axis-line')
                .data(allAxis);

            axisLines
                .enter()
                .append('line')
                .attr('class', 'axis-line')
                .style("stroke", "grey")
                .style("stroke-width", "1px");

            axisLines.exit().remove();

            g_axisLines.selectAll('.axis-line')
                .attr("x1", availableWidth / 2)
                .attr("y1", availableHeight / 2)
                .attr("x2", function (d, i) { return availableWidth / 2 + radius * factor * Math.sin(i * radians / allAxis.length); })
                .attr("y2", function (d, i) { return availableHeight / 2 - radius * factor * Math.cos(i * radians / allAxis.length); });

            // axis labels
            gEnter.append('g').attr('class', 'nv-radarAxisLabels');
            var g_axisLabels = wrap.select('.nv-radarAxisLabels');

            var axisLabels = g_axisLabels
                .selectAll('.axis-label')
                .data(allAxis);

            axisLabels
                .enter()
                .append('text')
                .attr('class', 'axis-label')
                .text(function (d) { return d; })
                .attr("dy", "1.5em")
                .attr("transform", function(d, i){return "translate(0, -10)"})
                .attr("text-anchor", "middle");

            axisLabels.exit().remove();

            g_axisLabels.selectAll('.axis-label')
                .attr("x", function (d, i) { return availableWidth / 2 + radius * Math.sin(i * radians / allAxis.length) + (this.getBBox().width / 1.5) * Math.sin(i * radians / allAxis.length); })
                .attr("y", function (d, i) { return availableHeight / 2 - radius * Math.cos(i * radians / allAxis.length) - 20 * Math.cos(i * radians / allAxis.length); });

            // series
            gEnter.append('g').attr('class', 'nv-radarSeries');
            var g_series = wrap.select('.nv-radarSeries');

            var series = g_series
                .selectAll('.series')
                .data(data);

            series
                .enter()
                .append('g')
                .attr('class', 'series');

            series.exit().remove();

            function getSeriesNodeData(d) {
                return d.values.map(function (v, i) {
                    var delta = (v.value - min) / (adjustedMax - min);

                    return {
                        x: availableWidth / 2 + radius * delta * factor * Math.sin(i * radians / allAxis.length),
                        y: availableHeight / 2 - radius * delta * factor * Math.cos(i * radians / allAxis.length),
                        series: d.series,
                        color: d.color,
                        value: v.value,
                        key: v.axis,
                        seriesName: d.key,
                        link: v.link || null
                    };
                });
            };

            // series areas
            var seriesAreas = series.selectAll('.series-area')
                .data(function (d) { return [getSeriesNodeData(d)]; });

            seriesAreas
                .enter()
                .append('polygon')
                .attr('class', function (d) { return 'series-area series-area-' + d[0].series; })
                .style('fill-opacity', opacityArea)
                .style('stroke-width', '2px')
                .on('mouseover', function (d){
                    var active = this;

                    series.selectAll(".series-area")
                        .filter(function () { return this != active; })
                        .transition(200)
                        .style("fill-opacity", 0.1);

                    d3.select(active)
                        .transition(200)
                        .style("fill-opacity", .7);
                })
                .on('mouseout', function(){
                    series.selectAll(".series-area")
                        .transition(200)
                        .style("fill-opacity", opacityArea);
                });

            seriesAreas.exit().remove();

            series.selectAll('.series-area')
                .attr('points', function (d) {
                    return d.map(function (o) { return o.x + ',' + o.y; }).join(' ');
                })
                .style('stroke', function (d) { return d[0].color; })
                .style('fill', function (d) { return d[0].color; });

            // series points
            var seriesPoints = series.selectAll('.series-point')
                .data(getSeriesNodeData);

            seriesPoints
                .enter()
                .append('circle')
                .attr('class', 'series-point')
                .attr('cursor', function (d) { return d.link ? 'pointer' : 'default'; })
                .attr('r', nodeRadius)
                .on('mouseover', function (d) {
                    d3.select(this).classed('hover', true);

                    dispatch.elementMouseover({
                        data: d
                    });
                })
                .on('mouseout', function(d) {
                    d3.select(this).classed('hover', false);

                    dispatch.elementMouseout({
                        data: d
                    });
                })
                .on('mousemove', function(d) {
                    dispatch.elementMousemove({
                        data: d
                    });
                })
                .on('click', function(d, i) {
                    dispatch.chartClick({
                        data: d,
                        index: i,
                        pos: d3.event,
                        id: id
                    });
                });

            seriesPoints.exit().remove();

            series.selectAll('.series-point')
                .attr('cx', function (d) { return d.x; })
                .attr('cy', function (d) { return d.y; })
                .style('fill', function (d) { return d.color; });
        });

        renderWatch.renderEnd('radar immediate');
        return chart;
    }

    //============================================================
    // Expose Public Variables
    //------------------------------------------------------------

    chart.dispatch = dispatch;
    chart.options = nv.utils.optionsFunc.bind(chart);

    chart._options = Object.create({}, {
        // simple options, just get/set the necessary values
        width:      {get: function(){return width;}, set: function(_){width=_;}},
        height:     {get: function(){return height;}, set: function(_){height=_;}},
        title:      {get: function(){return title;}, set: function(_){title=_;}},
        valueFormat:    {get: function(){return valueFormat;}, set: function(_){valueFormat=_;}},
        id:         {get: function(){return id;}, set: function(_){id=_;}},
        min:         {get: function(){return min;}, set: function(_){min=_;}},
        max: { get: function () { return max; }, set: function (_) { max = _; } },
        radius: { get: function () { return radius; }, set: function (_) { radius = _; } },
        factor: { get: function () { return factor; }, set: function (_) { factor = _; } },
        factorLegend: { get: function () { return factorLegend; }, set: function (_) { factorLegend = _; } },
        levels: { get: function () { return levels; }, set: function (_) { levels = _; } },
        opacityArea: { get: function () { return opacityArea; }, set: function (_) { opacityArea = _; } },

        // options that require extra logic in the setter
        margin: {get: function(){return margin;}, set: function(_){
            margin.top    = typeof _.top    != 'undefined' ? _.top    : margin.top;
            margin.right  = typeof _.right  != 'undefined' ? _.right  : margin.right;
            margin.bottom = typeof _.bottom != 'undefined' ? _.bottom : margin.bottom;
            margin.left   = typeof _.left   != 'undefined' ? _.left   : margin.left;
        }},
        color: {get: function(){return color;}, set: function(_){
            color=nv.utils.getColor(_);
        }}
    });

    nv.utils.initOptions(chart);
    return chart;
};